grafana-cli: Fix file path processing, returning of errors

aknuds1 opened the pull request:
**What this PR does / why we need it**:
Fix grafana-cli so it uses `filepath.Join` for filesystem path joining, so it works on Windows. Also fix some returning of errors.

aknuds1 committed:
grafana-cli: Fix file path processing, returning of errors

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

papagian reviewed (approved):
LGTM :heart:

I wonder how we haven't encountered any issues on Windows earlier. I guess not a lot of users use customised plugins in Windows?

papagian commented on pkg/cmd/grafana-cli/commands/install_command.go:
```diff
@@ -225,27 +224,26 @@ func extractFiles(archiveFile string, pluginName string, filePath string, allowS
 			} else {
-				err = extractFile(zf, newFile)
-				if err != nil {
+				if err := extractFile(zf, newFile); err != nil {
```
lint prefers `else if` here

aknuds1 replied:
@papagian Ah argh.

aknuds1 replied:
Will fix.

aknuds1 commented:
> I wonder how we haven't encountered any issues on Windows earlier. I guess not a lot of users use customised plugins in Windows?

@papagian Wondering the same...

aknuds1 committed:
Chore: Fix tests for Go 1.15

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Fix linting issues

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Fix tests

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

adrianlzt committed:
grafana-cli: Fix installing of plugins missing directory entries in zip (#26945)

aknuds1 committed:
Fix tests

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

kaydelaney committed:
Panel/BarGauge: Prevent overflow in panel with many series (#26926)

* Panel/BarGauge: Prevent overflow in panel with many series
Closes #24889

kaydelaney committed:
Datasource/CloudWatchLogs: Correctly interpolate variables for data links (#26929)

Closes #26792

kylebrandt committed:
Dataframe/Alerting: to tsdb.TimeSeriesSlice - accept "empty" time series (#26903)

this is in order to trigger no data when used with alerting.
fixes #26897.

This is perhaps better solved in the SDK but that needs more thought and is tracked in grafana/grafana-plugin-sdk-go#214

martinrw committed:
Docs: Mention httpMode in the docs (#26964)

hebbet committed:
Docs: Remove "currently in beta" (#26960)

7.1 is stable

NiteshKoushik committed:
Docs: Adding reference for the default database (#26950)

Adding reference for the default database type SQLite

lucasra1 committed:
Templating: Fixes renaming a variable using special characters or same name (#26866)

* Fix variable editor name-input bug

You couldn't delete an invalid character after typing it into the name-input field.
While investigating the issue turned out to be bigger, as there was a problem with
valid characters too. (See test scenarios below)

The fix seems to be, to remove an unnecessary check in the `changeVariableName`
action. There is theoretically now the possibility, that the `changeVariableName`
action is called with the same name, as the variable is already, but practically
there seems no possibility, that this could happen. A test, which checks that, had
to be removed too.

Test scenarios:
* 1st Scenario
    1. Type "@"
    2. Try deleting it
* 2nd Scenario
    1. Type "w"
    2. delete "w"
    3. Try typing "w" again

Fixes #26562

* Fix bug when updating existing variable

aknuds1 committed:
grafana-cli: Fix zip extraction

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
grafana-cli: Fix zip extraction

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Improve tests

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into chore/fix-cli-paths

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

hugohaggmark commented:
> > I wonder how we haven't encountered any issues on Windows earlier. I guess not a lot of users use customised plugins in Windows?
> 
> @papagian Wondering the same...

What was the issue on Windows?

papagian commented:
> What was the issue on Windows?

Assuming `filePath` is `C:\grafana\data\plugins\plugin` and `newFile` is `something` then `path.Join(filePath, newFileName)` becomes `C:\grafana\data\plugins\plugin/something` which is not valid path for Windows. Check [here](https://stackoverflow.com/questions/39181790/whats-the-difference-between-path-path-filepath-packages-in-go#answer-39182128) for more details.

aknuds1 commented:
@hugohaggmark @papagian is correct. We can't use `path.Join` for filesystem paths, since it assumes the path separator is '/'. The `path` package is meant for URI paths, not filesystem ones.

hugohaggmark commented:
Thank you @papagian @aknuds1 🙏
